README.md
Patchwork UTF-8 for PHP
=======================

[![Latest Stable Version](https://poser.pugx.org/patchwork/utf8/v/stable.png)](https://packagist.org/packages/patchwork/utf8)
[![Total Downloads](https://poser.pugx.org/patchwork/utf8/downloads.png)](https://packagist.org/packages/patchwork/utf8)
[![Build Status](https://secure.travis-ci.org/tchwork/utf8.png?branch=master)](http://travis-ci.org/tchwork/utf8)
[![SensioLabsInsight](https://insight.sensiolabs.com/projects/666c8ae7-0997-4d27-883a-6089ce3cc76b/mini.png)](https://insight.sensiolabs.com/projects/666c8ae7-0997-4d27-883a-6089ce3cc76b)

Patchwork UTF-8 gives PHP developpers extensive, portable and performant
handling of UTF-8 and [grapheme clusters](http://unicode.org/reports/tr29/).

It provides both :

- a portability layer for `mbstring`, `iconv`, and intl `Normalizer` and
  `grapheme_*` functions,
- an UTF-8 grapheme clusters aware replica of native string functions.

It can also serve as a documentation source referencing the practical problems
that arise when handling UTF-8 in PHP: Unicode concepts, related algorithms,
bugs in PHP core, workarounds, etc.

Version 1.2 adds best-fit mappings for UTF-8 to *Code Page* approximations.
It also adds Unicode filesystem access under Windows, using preferably
[wfio](https://github.com/kenjiuno/php-wfio) or a COM based fallback otherwise.

Portability
-----------

Unicode handling in PHP is best performed using a combo of `mbstring`, `iconv`,
`intl` and `pcre` with the `u` flag enabled. But when an application is expected
to run on many servers, you should be aware that these 4 extensions are not
always enabled.

Patchwork UTF-8 provides pure PHP implementations for 3 of those 4 extensions.
`pcre` compiled with unicode support is required but is widely available.
The following set of portability-fallbacks allows an application to run on a
server even if one or more of those extensions are not enabled:

- *utf8_encode, utf8_decode*,
- `mbstring`: *mb_check_encoding, mb_convert_case, mb_convert_encoding,
  mb_decode_mimeheader, mb_detect_encoding, mb_detect_order,
  mb_encode_mimeheader, mb_encoding_aliases, mb_internal_encoding, mb_language,
  mb_list_encodings, mb_strlen, mb_strpos, mb_strrpos, mb_strtolower,
  mb_strtoupper, mb_stripos, mb_stristr, mb_strrchr, mb_strrichr, mb_strripos,
  mb_strstr, mb_substitute_character, mb_substr*,
- `iconv`: *iconv, iconv_mime_decode, iconv_mime_decode_headers,
  iconv_get_encoding, iconv_set_encoding, iconv_mime_encode, ob_iconv_handler,
  iconv_strlen, iconv_strpos, iconv_strrpos, iconv_substr*,
- `intl`: *Normalizer, grapheme_extract, grapheme_stripos, grapheme_stristr,
  grapheme_strlen, grapheme_strpos, grapheme_strripos, grapheme_strrpos,
  grapheme_strstr, grapheme_substr, normalizer_is_normalized,
  normalizer_normalize*.

Patchwork\Utf8
--------------

[Grapheme clusters](http://unicode.org/reports/tr29/) should always be
considered when working with generic Unicode strings. The `Patchwork\Utf8`
class implements the quasi-complete set of native string functions that need
UTF-8 grapheme clusters awareness. Function names, arguments and behavior
carefully replicates native PHP string functions.

Some more functions are also provided to help handling UTF-8 strings:

- *filter()*: normalizes to UTF-8 NFC, converting from [CP-1252](http://wikipedia.org/wiki/CP-1252) when needed,
- *isUtf8()*: checks if a string contains well formed UTF-8 data,
- *toAscii()*: generic UTF-8 to ASCII transliteration,
- *strtocasefold()*: unicode transformation for caseless matching,
- *strtonatfold()*: generic case sensitive transformation for collation matching
- *wrapPath()*: unicode filesystem access under Windows and other OSes.

Mirrored string functions are:
*strlen, substr, strpos, stripos, strrpos, strripos, strstr, stristr, strrchr,
strrichr, strtolower, strtoupper, wordwrap, chr, count_chars, ltrim, ord, rtrim,
trim, str_ireplace, str_pad, str_shuffle, str_split, str_word_count, strcmp,
strnatcmp, strcasecmp, strnatcasecmp, strncasecmp, strncmp, strcspn, strpbrk,
strrev, strspn, strtr, substr_compare, substr_count, substr_replace, ucfirst,
lcfirst, ucwords, number_format, utf8_encode, utf8_decode, json_decode,
filter_input, filter_input_array*.

Notably missing (but hard to replicate) are *printf*-family functions.

The implementation favors performance over full edge cases handling.
It generally works on UTF-8 normalized strings and provides filters to get them.

As the turkish locale requires special cares, a `Patchwork\TurkishUtf8` class
is provided for working with this locale. It clones all the features of
`Patchwork\Utf8` but knows about the turkish specifics.

Usage
-----

The recommended way to install Patchwork UTF-8 is [through
composer](http://getcomposer.org). Just create a `composer.json` file and run
the `php composer.phar install` command to install it:

    {
        "require": {
            "patchwork/utf8": "~1.2"
        }
    }

Then, early in your bootstrap sequence, you have to configure your environment:

```php
\Patchwork\Utf8\Bootup::initAll(); // Enables the portablity layer and configures PHP for UTF-8
\Patchwork\Utf8\Bootup::filterRequestUri(); // Redirects to an UTF-8 encoded URL if it's not already the case
\Patchwork\Utf8\Bootup::filterRequestInputs(); // Normalizes HTTP inputs to UTF-8 NFC
```

Run `phpunit` to see the code in action.

Make sure that you are confident about using UTF-8 by reading
[Character Sets / Character Encoding Issues](http://www.phpwact.org/php/i18n/charsets)
and [Handling UTF-8 with PHP](http://www.phpwact.org/php/i18n/utf-8),
or [PHP et UTF-8](http://julp.lescigales.org/articles/3-php-et-utf-8.html) for french readers.

You should also get familiar with the concept of
[Unicode Normalization](http://en.wikipedia.org/wiki/Unicode_equivalence) and
[Grapheme Clusters](http://unicode.org/reports/tr29/).

Do not blindly replace all use of PHP's string functions. Most of the time you
will not need to, and you will be introducing a significant performance overhead
to your application.

Screen your input on the *outer perimeter* so that only well formed UTF-8 pass
through. When dealing with badly formed UTF-8, you should not try to fix it
(see [Unicode Security Considerations](http://www.unicode.org/reports/tr36/#Deletion_of_Noncharacters)).
Instead, consider it as [CP-1252](http://wikipedia.org/wiki/CP-1252) and use
`Patchwork\Utf8::utf8_encode()` to get an UTF-8 string. Don't forget also to
choose one unicode normalization form and stick to it. NFC is now the defacto
standard. `Patchwork\Utf8::filter()` implements this behavior: it converts from
CP1252 and to NFC.

This library is orthogonal to `mbstring.func_overload` and will not work if the
php.ini setting is enabled.

Licensing
---------

Patchwork\Utf8 is free software; you can redistribute it and/or modify it under
the terms of the (at your option):
- [Apache License v2.0](http://apache.org/licenses/LICENSE-2.0.txt), or
- [GNU General Public License v2.0](http://gnu.org/licenses/gpl-2.0.txt).

Unicode handling requires tedious work to be implemented and maintained on the
long run. As such, contributions such as unit tests, bug reports, comments or
patches licensed under both licenses are really welcomed.

I hope many projects could adopt this code and together help solve the unicode
subject for PHP.


CHANGELOG.md
## v1.3.0 (2015-12-15)

- add shim for mb_convert_variables
- marked all shims as @internal
- test on appveyor
- a few fixes in iconv and mbstring shims
- cleanup refacto for preparing v2 based on symfony-polyfill

## v1.2.5 (2015-10-14)

- handle the third argument of mb_convert_encoding() being an array
- add license files

## v1.2.4 (2015-06-29)

- trigger silenced deprecation notices when shims are in use
- fix mb_strrpos() shim registration
- fix .gitattributes

## v1.2.3 (2015-06-25)

- fix mb_strrpos shim with negative offset
- sync tests with latest PHP/HHVM behaviors
- remove PHP7/HHVM from allowed failures
- move to PSR-1+2+4
- mv class/ src/

## v1.2.2 (2015-04-26)

- Fix ucwords to be functionally the same as in-built PHP version
- Fix iconv_set_encoding deprecation notice in PHP 5.6.0
- remove legacy test for HHVM/PHP7
- mb_parse_str() should have no return value

## v1.2.1 (2015-01-28)

- fix double declaration in mbstring shim

## v1.2.0 (2015-01-12)

- add u::strwidth() to get the width of a string when printed on a terminal
- add more mbstring shims
- add a note about https://bugs.php.net/65358
- fail properly when COM is not loaded
- fallback on stat() when lstat() fails

## v1.2.0-beta (2014-08-05)

- add best-fit mappings for UTF-8 to Code Page approximations
- add portable Unicode filesystem access under Windows and other OSes

## v1.1.29 (2015-04-26)

- fix ucwords to be functionally the same as in-built PHP version
- fix iconv_set_encoding deprecation notice in PHP 5.6.0
- remove legacy test for HHVM/PHP7

## v1.1.28 (2015-01-12)

- fix mbstring shim for html-entities

## v1.1.27 (2015-01-11)

- update to Unicode 7.0
- fix iconv shim compat layer

## v1.1.26 (2014-11-08)

- tweak composer.json

## v1.1.25 (2014-08-05)

- update travis matrix
- add composer branch alias

## v1.1.24 (2014-06-17)

- update tests for latest HHVM fixes
- move legacy GRAPHEME_CLUSTER_RX version to Intl shim

## v1.1.23 (2014-05-22)

- enable tests for PHP 5.6
- remove HHVM from allowed failures

## v1.1.22 (2014-05-06)

- fix #19: don't call ini_set() when not required and gain compat with PHP5.6

## v1.1.21 (2014-03-26)

- fix #18 u::wordwrap() now relies on native behavior

## v1.1.20 (2014-03-01)
## v1.1.19 (2014-03-01)

- fix mb_regex_encoding() being disabled on some hosting providers

## v1.1.18 (2014-02-02)

- require PCRE>=7.3, the first that correctly checks UTF-8 validity
- enable HHVM on Travis CI

## v1.1.17 (2014-01-02)

- enable Travis CI and SensioLabsInsight
- add shims for mb_check_encoding, mb_detect_encoding, mb_detect_order,
  mb_language and mb_encoding_aliases
- mbstring shim fix: alias UTF8 to UTF-8
- more tests

## v1.1.16 (2013-12-06)

- fix $_FILES bootup filtering
- fix mbstring shim behavior with invalid utf8 strings

## v1.1.15 (2013-11-23)

- u::toAscii() is now locale sensitive and allows a substitution character
- use LSB for more extension openness
- handle null for mb_substr() shim length as in PHP 5.4.8
- fix casts to string
- fix mbstring MB_CASE_TITLE shim on edge case
- small optimizations
- add a changelog

## v1.1.14 (2013-11-04)

- set default_charset to UTF-8 at bootup
- remove bootup PCRE warning
- fix iconv internal_encoding shim
- fix bootup dependencies
- add tests for normalizers consts
- readme update

## v1.1.13 (2013-10-11)

- new u::filter(): normalizes to UTF-8 NFC, converting from CP-1252 when needed
- new u::json_decode(), u::filter_input() and u::filter_input_array() for NFC safeness
- reference Unicode 6.3
- more tests
- readme update

## v1.1.12 (2013-10-03)

- new Patchwork\TurkishUtf8 class extends Patchwork\Utf8 with Turkish specifics
- expose Patchwork\Utf8\Bootup::filterString() for UF-8 NFC strings normalization
- normalize inputs EOL to work around https://bugs.php.net/65732
- update composer.json

## v1.1.11 (2013-08-19)

- updates related to PHP bugs 52211 and 61860
- fixes and tests for iconv shim
- fixes and tests for mbstring shim

## v1.1.10 (2013-08-13)

- update .gitattributes export-ignore
- fixes and tests for intl::grapheme_extract() shim
- fixes and tests for iconv shim
- fixes and tests for mbstring shim

## v1.1.9 (2013-08-04)

- know that PHP bug 61860 has been fixed in 5.5.1
- fix intl::grapheme_strlen() shim on edge case
- fix case sensitive encoding checks for mbstring shim
- some more fixes, tests and optimizations

## v1.1.8 since v1.1.0 (2013-05-24)

- filter leading combining chars in inputs for NFC safeness
- fixes, tests and optimizations
- readme update

## v1.1.0 (2013-04-18)

- PSR-0 autoloading and explicit bootup configuration is now required

## v1.0.6 since v1.0.0 (2013-04-22)

- add extra characters for ASCII transliterations
- move bootup stages in namespaced functions for greater modularity
- NFC normalization for autoglobal inputs
- better setlocale() initialization

class/Patchwork/Utf8.php
<?php // vi: set fenc=utf-8 ts=4 sw=4 et:
/*
 * Copyright (C) 2013 Nicolas Grekas - p@tchwork.com
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the (at your option):
 * Apache License v2.0 (http://apache.org/licenses/LICENSE-2.0.txt), or
 * GNU General Public License v2.0 (http://gnu.org/licenses/gpl-2.0.txt).
 */

namespace Patchwork;

use Normalizer as n;

/**
 * UTF-8 Grapheme Cluster aware string manipulations implementing the quasi complete
 * set of native PHP string functions that need UTF-8 awareness and more.
 * Missing are printf-family functions.
 */
class Utf8
{
    protected static

    $pathPrefix,
    $commonCaseFold = array(
        array('µ','ſ',"\xCD\x85",'ς',"\xCF\x90","\xCF\x91","\xCF\x95","\xCF\x96","\xCF\xB0","\xCF\xB1","\xCF\xB5","\xE1\xBA\x9B","\xE1\xBE\xBE"),
        array('μ','s','ι',       'σ','β',       'θ',       'φ',       'π',       'κ',       'ρ',       'ε',       "\xE1\xB9\xA1",'ι'           )
    ),
    $cp1252 = array('','','','','','','','','','','','','','','','','','','','','','','','','','',''),
    $utf8   = array('€','‚','ƒ','„','…','†','‡','ˆ','‰','Š','‹','Œ','Ž','‘','’','“','”','•','–','—','˜','™','š','›','œ','ž','Ÿ');


    static function isUtf8($s)
    {
        return (bool) preg_match('//u', $s); // Since PHP 5.2.5, this also excludes invalid five and six bytes sequences
    }

    // Generic UTF-8 to ASCII transliteration

    static function toAscii($s, $subst_chr = '?')
    {
        if (preg_match("/[\x80-\xFF]/", $s))
        {
            static $translitExtra = array();
            $translitExtra or $translitExtra = static::getData('translit_extra');

            $s = n::normalize($s, n::NFKC);

/**/        $glibc = 'glibc' === ICONV_IMPL;

            preg_match_all('/./u', $s, $s);

            foreach ($s[0] as &$c)
            {
                if (! isset($c[1])) continue;

/**/            if ($glibc)
/**/            {
                    $t = iconv('UTF-8', 'ASCII//TRANSLIT', $c);
/**/            }
/**/            else
/**/            {
                    $t = iconv('UTF-8', 'ASCII//IGNORE//TRANSLIT', $c);

                    if (! isset($t[0])) $t = '?';
                    else if (isset($t[1])) $t = ltrim($t, '\'`"^~');
/**/            }

                if ('?' === $t)
                {
                    if (isset($translitExtra[$c]))
                    {
                        $t = $translitExtra[$c];
                    }
                    else
                    {
                        $t = n::normalize($c, n::NFD);

                        if ($t[0] < "\x80") $t = $t[0];
                        else $t = $subst_chr;
                    }
                }

                $c = $t;
            }

            $s = implode('', $s[0]);
        }

        return $s;
    }

    static function wrapPath($path = '')
    {
        if (null === static::$pathPrefix)
        {
            if (extension_loaded('wfio'))
            {
                static::$pathPrefix = 'wfio://';
            }
            else if ('\\' === DIRECTORY_SEPARATOR && class_exists('COM', false))
            {
                static::$pathPrefix = 'utf8'.mt_rand();
                stream_wrapper_register(static::$pathPrefix, 'Patchwork\Utf8\WindowsStreamWrapper');
                static::$pathPrefix .= '://';
            } else {
                if ('\\' === DIRECTORY_SEPARATOR) {
                    trigger_error('The `wfio` or `com_dotnet` extension is required to handle UTF-8 filesystem access on Windows');
                }
                static::$pathPrefix = 'file://';
            }
        }

        return static::$pathPrefix . $path;
    }

    static function filter($var, $normalization_form = 4 /* n::NFC */, $leading_combining = '◌')
    {
        switch (gettype($var))
        {
        case 'array':
            foreach ($var as $k => $v) $var[$k] = static::filter($v, $normalization_form, $leading_combining);
            break;

        case 'object':
            foreach ($var as $k => $v) $var->$k = static::filter($v, $normalization_form, $leading_combining);
            break;

        case 'string':
            if (false !== strpos($var, "\r"))
            {
                // Workaround https://bugs.php.net/65732
                $var = str_replace("\r\n", "\n", $var);
                $var = strtr($var, "\r", "\n");
            }

            if (preg_match('/[\x80-\xFF]/', $var))
            {
                if (n::isNormalized($var, $normalization_form)) $n = '-';
                else
                {
                    $n = n::normalize($var, $normalization_form);
                    if (isset($n[0])) $var = $n;
                    else $var = static::utf8_encode($var);
                }

                if ($var[0] >= "\x80" && isset($n[0], $leading_combining[0]) && preg_match('/^\p{Mn}/u', $var))
                {
                    // Prevent leading combining chars
                    // for NFC-safe concatenations.
                    $var = $leading_combining . $var;
                }
            }
            break;
        }

        return $var;
    }

    // Unicode transformation for caseless matching
    // see http://unicode.org/reports/tr21/tr21-5.html

    static function strtocasefold($s, $full = true)
    {
        $s = str_replace(self::$commonCaseFold[0], self::$commonCaseFold[1], $s);

        if ($full)
        {
            static $fullCaseFold = false;
            $fullCaseFold || $fullCaseFold = static::getData('caseFolding_full');

            $s = str_replace($fullCaseFold[0], $fullCaseFold[1], $s);
        }

        return static::strtolower($s);
    }

    // Generic case sensitive collation support for self::strnatcmp()

    static function strtonatfold($s)
    {
        $s = n::normalize($s, n::NFD);
        return preg_replace('/\p{Mn}+/u', '', $s);
    }

    // PHP string functions that need UTF-8 awareness

    static function filter_input($type, $var, $filter = FILTER_DEFAULT, $option = null)
    {
        if (4 > func_num_args()) $var = filter_input($type, $var, $filter);
        else $var = filter_input($type, $var, $filter, $option);

        return static::filter($var);
    }

    static function filter_input_array($type, $def = null, $add_empty = true)
    {
        if (2 > func_num_args()) $a = filter_input_array($type);
        else $a = filter_input_array($type, $def, $add_empty);

        return static::filter($a);
    }

    static function json_decode($json, $assoc = false, $depth = 512, $options = 0)
    {
/**/    if (PHP_VERSION_ID < 50400)
/**/    {
            $json = json_decode($json, $assoc, $depth);
/**/    }
/**/    else
/**/    {
            $json = json_decode($json, $assoc, $depth, $options);
/**/    }

        return static::filter($json);
    }

    static function substr($s, $start, $len = 2147483647)
    {
/**/    static $bug62759;

/**/    isset($bug62759) or $bug62759 = extension_loaded('intl') && 'à' === grapheme_substr('éà', 1, -2);

/**/    if ($bug62759)
/**/    {
            return PHP\Shim\Intl::grapheme_substr_workaround62759($s, $start, $len);
/**/    }
/**/    else
/**/    {
            return grapheme_substr($s, $start, $len);
/**/    }
    }

    static function strlen($s) {return grapheme_strlen($s);}
    static function strpos ($s, $needle, $offset = 0) {return grapheme_strpos ($s, $needle, $offset);}
    static function strrpos($s, $needle, $offset = 0) {return grapheme_strrpos($s, $needle, $offset);}

    static function stripos($s, $needle, $offset = 0)
    {
/**/    if (50418 > PHP_VERSION_ID || 50500 == PHP_VERSION_ID)
/**/    {
            // Don't use grapheme_stripos because of https://bugs.php.net/61860
            if (! preg_match('//u', $s .= '')) return false;
            if ($offset < 0) $offset = 0;
            if (! $needle = mb_stripos($s, $needle .= '', $offset, 'UTF-8')) return $needle;
            return grapheme_strlen(iconv_substr($s, 0, $needle, 'UTF-8'));
/**/    }
/**/    else
/**/    {
            return grapheme_stripos($s, $needle, $offset);
/**/    }
    }

    static function strripos($s, $needle, $offset = 0)
    {
/**/    if (50418 > PHP_VERSION_ID || 50500 == PHP_VERSION_ID)
/**/    {
            // Don't use grapheme_strripos because of https://bugs.php.net/61860
            if (! preg_match('//u', $s .= '')) return false;
            if ($offset < 0) $offset = 0;
            if (! $needle = mb_strripos($s, $needle .= '', $offset, 'UTF-8')) return $needle;
            return grapheme_strlen(iconv_substr($s, 0, $needle, 'UTF-8'));
/**/    }
/**/    else
/**/    {
            return grapheme_strripos($s, $needle, $offset);
/**/    }
    }

    static function stristr($s, $needle, $before_needle = false)
    {
        if ('' === $needle .= '') return false;
        return mb_stristr($s, $needle, $before_needle, 'UTF-8');
    }

    static function strstr  ($s, $needle, $before_needle = false) {return grapheme_strstr($s, $needle, $before_needle);}
    static function strrchr ($s, $needle, $before_needle = false) {return mb_strrchr ($s, $needle, $before_needle, 'UTF-8');}
    static function strrichr($s, $needle, $before_needle = false) {return mb_strrichr($s, $needle, $before_needle, 'UTF-8');}

    static function strtolower($s) {return mb_strtolower($s, 'UTF-8');}
    static function strtoupper($s) {return mb_strtoupper($s, 'UTF-8');}

    static function wordwrap($s, $width = 75, $break = "\n", $cut = false)
    {
        if (false === wordwrap('-', $width, $break, $cut)) return false;

        is_string($break) or $break = (string) $break;

        $w = '';
        $s = explode($break, $s);
        $iLen = count($s);
        $chars = array();

        if (1 === $iLen && '' === $s[0])
            return '';

        for ($i = 0; $i < $iLen; ++$i)
        {
            if ($i)
            {
                $chars[] = $break;
                $w .= '#';
            }

            $c = $s[$i];
            unset($s[$i]);

            foreach (self::str_split($c) as $c)
            {
                $chars[] = $c;
                $w .= ' ' === $c ? ' ' : '?';
            }
        }

        $s = '';
        $j = 0;
        $b = $i = -1;
        $w = wordwrap($w, $width, '#', $cut);

        while (false !== $b = strpos($w, '#', $b+1))
        {
            for (++$i; $i < $b; ++$i)
            {
                $s .= $chars[$j];
                unset($chars[$j++]);
            }

            if ($break === $chars[$j] || ' ' === $chars[$j]) unset($chars[$j++]);
            $s .= $break;
        }

        return $s . implode('', $chars);
    }

    static function chr($c)
    {
        if (0x80 > $c %= 0x200000) return chr($c);
        if (0x800 > $c) return chr(0xC0 | $c>>6) . chr(0x80 | $c & 0x3F);
        if (0x10000 > $c) return chr(0xE0 | $c>>12) . chr(0x80 | $c>>6 & 0x3F) . chr(0x80 | $c & 0x3F);
        return chr(0xF0 | $c>>18) . chr(0x80 | $c>>12 & 0x3F) . chr(0x80 | $c>>6 & 0x3F) . chr(0x80 | $c & 0x3F);
    }

    static function count_chars($s, $mode = 0)
    {
        if (1 != $mode) user_error(__METHOD__ . '(): the only allowed $mode is 1', E_USER_WARNING);
        $s = self::str_split($s);
        return array_count_values($s);
    }

    static function ltrim($s, $charlist = INF)
    {
        $charlist = INF === $charlist ? '\s' : self::rxClass($charlist);
        return preg_replace("/^{$charlist}+/u", '', $s);
    }

    static function ord($s)
    {
        $a = ($s = unpack('C*', substr($s, 0, 4))) ? $s[1] : 0;
        if (0xF0 <= $a) return (($a - 0xF0)<<18) + (($s[2] - 0x80)<<12) + (($s[3] - 0x80)<<6) + $s[4] - 0x80;
        if (0xE0 <= $a) return (($a - 0xE0)<<12) + (($s[2] - 0x80)<<6) + $s[3] - 0x80;
        if (0xC0 <= $a) return (($a - 0xC0)<<6) + $s[2] - 0x80;
        return $a;
    }

    static function rtrim($s, $charlist = INF)
    {
        $charlist = INF === $charlist ? '\s' : self::rxClass($charlist);
        return preg_replace("/{$charlist}+$/u", '', $s);
    }

    static function trim($s, $charlist = INF) {return self::rtrim(self::ltrim($s, $charlist), $charlist);}

    static function str_ireplace($search, $replace, $subject, &$count = null)
    {
        $search = (array) $search;

        foreach ($search as $i => $s)
        {
            if ('' === $s .= '') $s = '/^(?<=.)$/';
            else $s = '/' . preg_quote($s, '/') . '/ui';

            $search[$i] = $s;
        }

        $subject = preg_replace($search, $replace, $subject, -1, $replace);
        $count = $replace;

        return $subject;
    }

    static function str_pad($s, $len, $pad = ' ', $type = STR_PAD_RIGHT)
    {
        $slen = grapheme_strlen($s);
        if ($len <= $slen) return $s;

        $padlen = grapheme_strlen($pad);
        $freelen = $len - $slen;
        $len = $freelen % $padlen;

        if (STR_PAD_RIGHT == $type) return $s . str_repeat($pad, $freelen / $padlen) . ($len ? grapheme_substr($pad, 0, $len) : '');
        if (STR_PAD_LEFT  == $type) return      str_repeat($pad, $freelen / $padlen) . ($len ? grapheme_substr($pad, 0, $len) : '') . $s;
        if (STR_PAD_BOTH  == $type)
        {
            $freelen /= 2;

            $type = ceil($freelen);
            $len = $type % $padlen;
            $s .= str_repeat($pad, $type / $padlen) . ($len ? grapheme_substr($pad, 0, $len) : '');

            $type = floor($freelen);
            $len = $type % $padlen;
            return str_repeat($pad, $type / $padlen) . ($len ? grapheme_substr($pad, 0, $len) : '') . $s;
        }

        user_error(__METHOD__ . '(): Padding type has to be STR_PAD_LEFT, STR_PAD_RIGHT, or STR_PAD_BOTH', E_USER_WARNING);
    }

    static function str_shuffle($s)
    {
        $s = self::str_split($s);
        shuffle($s);
        return implode('', $s);
    }

    static function str_split($s, $len = 1)
    {
        if (1 > $len = (int) $len)
        {
            $len = func_get_arg(1);
            return str_split($s, $len);
        }

/**/    if (extension_loaded('intl'))
/**/    {
            $a = array();
            $p = 0;
            $l = strlen($s);

            while ($p < $l) $a[] = grapheme_extract($s, 1, GRAPHEME_EXTR_COUNT, $p, $p);
/**/    }
/**/    else
/**/    {
            preg_match_all('/' . GRAPHEME_CLUSTER_RX . '/u', $s, $a);
            $a = $a[0];
/**/    }

        if (1 == $len) return $a;

        $s = array();
        $p = -1;

        foreach ($a as $l => $a)
        {
            if ($l % $len) $s[$p] .= $a;
            else $s[++$p] = $a;
        }

        return $s;
    }

    static function str_word_count($s, $format = 0, $charlist = '')
    {
        $charlist = self::rxClass($charlist, '\pL');
        $s = preg_split("/({$charlist}+(?:[\p{Pd}’']{$charlist}+)*)/u", $s, -1, PREG_SPLIT_DELIM_CAPTURE);

        $charlist = array();
        $len = count($s);

        if (1 == $format) for ($i = 1; $i < $len; $i+=2) $charlist[] = $s[$i];
        else if (2 == $format)
        {
            $offset = grapheme_strlen($s[0]);
            for ($i = 1; $i < $len; $i+=2)
            {
                $charlist[$offset] = $s[$i];
                $offset += grapheme_strlen($s[$i]) + grapheme_strlen($s[$i+1]);
            }
        }
        else $charlist = ($len - 1) / 2;

        return $charlist;
    }

    static function strcmp       ($a, $b) {return $a . '' === $b . '' ? 0 : strcmp(n::normalize($a, n::NFD), n::normalize($b, n::NFD));}
    static function strnatcmp    ($a, $b) {return $a . '' === $b . '' ? 0 : strnatcmp(self::strtonatfold($a), self::strtonatfold($b));}
    static function strcasecmp   ($a, $b) {return self::strcmp   (static::strtocasefold($a), static::strtocasefold($b));}
    static function strnatcasecmp($a, $b) {return self::strnatcmp(static::strtocasefold($a), static::strtocasefold($b));}
    static function strncasecmp  ($a, $b, $len) {return self::strncmp(static::strtocasefold($a), static::strtocasefold($b), $len);}
    static function strncmp      ($a, $b, $len) {return self::strcmp(self::substr($a, 0, $len), self::substr($b, 0, $len));}

    static function strcspn($s, $charlist, $start = 0, $len = 2147483647)
    {
        if ('' === $charlist .= '') return null;
        if ($start || 2147483647 != $len) $s = self::substr($s, $start, $len);

        return preg_match('/^(.*?)' . self::rxClass($charlist) . '/us', $s, $len) ? grapheme_strlen($len[1]) : grapheme_strlen($s);
    }

    static function strpbrk($s, $charlist)
    {
        if (preg_match('/' . self::rxClass($charlist) . '/us', $s, $m)) return substr($s, strpos($s, $m[0]));
        else return false;
    }

    static function strrev($s)
    {
        $s = self::str_split($s);
        return implode('', array_reverse($s));
    }

    static function strspn($s, $mask, $start = 0, $len = 2147483647)
    {
        if ($start || 2147483647 != $len) $s = self::substr($s, $start, $len);
        return preg_match('/^' . self::rxClass($mask) . '+/u', $s, $s) ? grapheme_strlen($s[0]) : 0;
    }

    static function strtr($s, $from, $to = INF)
    {
        if (INF !== $to)
        {
            $from = self::str_split($from);
            $to   = self::str_split($to);

            $a = count($from);
            $b = count($to);

                 if ($a > $b) $from = array_slice($from, 0, $b);
            else if ($a < $b) $to   = array_slice($to  , 0, $a);

            $from = array_combine($from, $to);
        }

        return strtr($s, $from);
    }

    static function substr_compare($a, $b, $offset, $len = 2147483647, $i = 0)
    {
        $a = self::substr($a, $offset, $len);
        return $i ? static::strcasecmp($a, $b) : self::strcmp($a, $b);
    }

    static function substr_count($s, $needle, $offset = 0, $len = 2147483647)
    {
        return substr_count(self::substr($s, $offset, $len), $needle);
    }

    static function substr_replace($s, $replace, $start, $len = 2147483647)
    {
        $s       = self::str_split($s);
        $replace = self::str_split($replace);
        array_splice($s, $start, $len, $replace);
        return implode('', $s);
    }

    static function ucfirst($s)
    {
        $c = iconv_substr($s, 0, 1, 'UTF-8');
        return static::strtoupper($c) . substr($s, strlen($c));
    }

    static function lcfirst($s)
    {
        $c = iconv_substr($s, 0, 1, 'UTF-8');
        return static::strtolower($c) . substr($s, strlen($c));
    }

    static function ucwords($s)
    {
        return preg_replace_callback(
            "/\b(.)/u",
            function ($matches) {
                return mb_convert_case($matches[1], MB_CASE_UPPER, 'UTF-8');
            },
            $s
        );
    }

    static function number_format($number, $decimals = 0, $dec_point = '.', $thousands_sep = ',')
    {
/**/    if (PHP_VERSION_ID < 50400)
/**/    {
            if (isset($thousands_sep[1]) || isset($dec_point[1]))
            {
                return str_replace(
                    array('.', ','),
                    array($dec_point, $thousands_sep),
                    number_format($number, $decimals, '.', ',')
                );
            }
/**/    }

        return number_format($number, $decimals, $dec_point, $thousands_sep);
    }

    static function utf8_encode($s)
    {
        $s = utf8_encode($s);
        if (false === strpos($s, "\xC2")) return $s;
        else return str_replace(self::$cp1252, self::$utf8, $s);
    }

    static function utf8_decode($s)
    {
        $s = str_replace(self::$utf8, self::$cp1252, $s);
        return utf8_decode($s);
    }

    static function strwidth($s)
    {
        if (false !== strpos($s, "\r"))
        {
            $s = str_replace("\r\n", "\n", $s);
            $s = strtr($s, "\r", "\n");
        }
        $width = 0;

        foreach (explode("\n", $s) as $s)
        {
            $s = preg_replace('/\x1B\[[\d;]*m/', '', $s);
            $c = substr_count($s, "\xAD") - substr_count($s, "\x08");
            $s = preg_replace('/[\x00\x05\x07\p{Mn}\p{Me}\p{Cf}\x{1160}-\x{11FF}\x{200B}]+/u', '', $s);
            preg_replace('/[\x{1100}-\x{115F}\x{2329}\x{232A}\x{2E80}-\x{303E}\x{3040}-\x{A4CF}\x{AC00}-\x{D7A3}\x{F900}-\x{FAFF}\x{FE10}-\x{FE19}\x{FE30}-\x{FE6F}\x{FF00}-\x{FF60}\x{FFE0}-\x{FFE6}\x{20000}-\x{2FFFD}\x{30000}-\x{3FFFD}]/u', '', $s, -1, $wide);

            if ($width < $c = iconv_strlen($s, 'UTF-8') + $wide + $c)
            {
                $width = $c;
            }
        }

        return $width;
    }

    protected static function rxClass($s, $class = '')
    {
        $class = array($class);

        foreach (self::str_split($s) as $s)
        {
            if ('-' === $s) $class[0] = '-' . $class[0];
            else if (!isset($s[2])) $class[0] .= preg_quote($s, '/');
            else if (1 === iconv_strlen($s, 'UTF-8')) $class[0] .= $s;
            else $class[] = $s;
        }

        $class[0] = '[' . $class[0] . ']';

        if (1 === count($class)) return $class[0];
        else return '(?:' . implode('|', $class) . ')';
    }

    protected static function getData($file)
    {
        $file = __DIR__ . '/Utf8/data/' . $file . '.ser';
        if (file_exists($file)) return unserialize(file_get_contents($file));
        else return false;
    }
}
